File: scripts/generate_demo_city.py
import json
import random
import math
import os
import sys

def generate_demo_city():
    """
    Generate demo city matching the backend LayoutEngine's output structure.

    Real repos analyzed by the backend typically produce:
    - Many districts (30-80), most with 1-3 files, a few with 4-6
    - District centers on golden-ratio spiral at spacing 25 * sqrt(i)
    - Buildings within districts on spiral at spacing 18 * sqrt(i)
    - Total city span ~400-500 units
    """

    # Realistic directory structure (each becomes a Leiden cluster / district)
    # Mirrors what the backend clustering produces: many small clusters
    district_templates = [
        # Core (few files each — typical isolated modules)
        ("src/core/auth", 3),
        ("src/core/db", 2),
        ("src/core/network", 4),
        ("src/core/crypto", 2),
        ("src/core/config", 1),
        ("src/core/logger", 1),
        ("src/core/events", 2),
        ("src/core/cache", 1),
        # API (small controllers)
        ("src/api/routes", 5),
        ("src/api/middleware", 3),
        ("src/api/controllers", 4),
        ("src/api/validators", 2),
        ("src/api/errors", 1),
        # UI (components tend to cluster in small groups)
        ("src/ui/components/layout", 3),
        ("src/ui/components/forms", 4),
        ("src/ui/components/cards", 2),
        ("src/ui/components/modals", 2),
        ("src/ui/hooks", 3),
        ("src/ui/styles", 1),
        ("src/ui/pages/home", 2),
        ("src/ui/pages/dashboard", 3),
        ("src/ui/pages/settings", 2),
        ("src/ui/pages/auth", 2),
        # Utils (mostly singletons)
        ("src/utils/math", 1),
        ("src/utils/format", 1),
        ("src/utils/validation", 2),
        ("src/utils/helpers", 1),
        ("src/utils/constants", 1),
        # Services
        ("src/services/user", 2),
        ("src/services/payment", 3),
        ("src/services/email", 1),
        ("src/services/analytics", 2),
        ("src/services/storage", 1),
        # Workers
        ("src/workers/queue", 2),
        ("src/workers/scheduler", 1),
        ("src/workers/cache", 1),
        # Tests (small clusters)
        ("tests/unit/core", 2),
        ("tests/unit/api", 3),
        ("tests/unit/ui", 2),
        ("tests/integration", 2),
        ("tests/e2e", 1),
        # Config / Root files (singletons typical in real repos)
        ("config", 1),
        ("scripts/deploy", 1),
        ("scripts/build", 1),
        ("docs", 1),
    ]

    languages = ['javascript', 'typescript', 'python', 'go', 'rust', 'css', 'json']
    authors = ['Yash Kamireddi', 'Alex Chen', 'Sarah Jenkins', 'Miguel Santos', 'Bot-Auto', 'Emma Watson']

    buildings = []
    edges = []

    file_id_counter = 1
    phi = (1 + math.sqrt(5)) / 2

    # Sort by size descending (largest cluster in center — matches backend)
    sorted_districts = sorted(district_templates, key=lambda d: d[1], reverse=True)

    # 1. Generate Buildings
    for d_idx, (district, num_buildings) in enumerate(sorted_districts):
        lang = random.choice(languages)

        # District center: golden-ratio spiral — exact same as backend LayoutEngine
        r = 35 * math.sqrt(d_idx)
        theta = d_idx * 2 * math.pi / (phi * phi)
        noise_r = random.uniform(-3, 3)
        noise_theta = random.uniform(-0.1, 0.1)
        cx = (r + noise_r) * math.cos(theta + noise_theta)
        cz = (r + noise_r) * math.sin(theta + noise_theta)

        for i in range(num_buildings):
            fid = f"b_{file_id_counter}"
            file_id_counter += 1

            # Building position: spiral within district — exact same as backend
            local_r = 24 * math.sqrt(i)
            local_theta = i * 2 * math.pi / (phi * phi)
            jitter = random.uniform(-1, 1)
            bx = cx + local_r * math.cos(local_theta) + jitter
            bz = cz + local_r * math.sin(local_theta) + jitter

            # Distribution: 5% massive monoliths, 15% mid-sized services, 80% tiny utilities
            roll = random.random()
            if roll < 0.05:
                complexity = random.randint(150, 400)
                deps_in = random.randint(30, 100)
                churn = random.randint(70, 100)
                age_days = random.randint(1000, 3000)
            elif roll < 0.20:
                complexity = random.randint(30, 100)
                deps_in = random.randint(10, 35)
                churn = random.randint(30, 80)
                age_days = random.randint(300, 1500)
            else:
                complexity = random.randint(2, 25)
                deps_in = random.randint(0, 8)
                churn = random.randint(0, 40)
                age_days = random.randint(10, 500)

            deps_out = random.randint(0, 8)

            # Dimensions — exponential scaling for epic skyline contrasts
            height = max(2, min(90, 4 + math.log2(1 + complexity) * 10))
            girth = max(3, min(20, 3 + math.log2(1 + deps_in) * 4))

            classes = []
            if lang in ['typescript', 'python', 'javascript']:
                for c in range(random.randint(0, 3)):
                    classes.append({"name": f"Class_{fid}_{c}", "line": random.randint(10, 100)})

            functions = []
            if lang in ['typescript', 'python', 'javascript', 'go', 'rust']:
                for f_idx in range(random.randint(1, 8)):
                    functions.append({"name": f"handle_{fid}_{f_idx}", "line": random.randint(10, 200)})

            exports = [f"export_{fid}_{x}" for x in range(random.randint(0, 4))]

            buildings.append({
                "id": fid,
                "name": f"Module_{fid}.{lang[:2]}",
                "path": f"{district}/Module_{fid}.{lang[:2]}",
                "district_id": f"district_{d_idx}",
                "position": {"x": round(bx, 1), "y": 0, "z": round(bz, 1)},
                "dimensions": {"width": round(girth, 1), "height": round(height, 1), "depth": round(girth, 1)},
                "metrics": {
                    "loc": random.randint(50, 1500),
                    "complexity": complexity,
                    "churn": random.randint(0, 100),
                    "age_days": random.randint(1, 1000),
                    "dependencies_in": deps_in,
                    "dependencies_out": deps_out,
                    "size_bytes": random.randint(1000, 50000)
                },
                "language": lang,
                "author": random.choice(authors),
                "email": "dev@codebase.city",
                "last_modified": "2023-10-25T14:32:00Z",
                "classes": classes,
                "functions": functions,
                "exports": exports
            })

    # 2. Generate Dependencies (Edges) — realistic dependency flow
    core_files = [b['id'] for b in buildings if 'core' in b['path']]
    api_files = [b['id'] for b in buildings if 'api' in b['path']]
    ui_files = [b['id'] for b in buildings if 'ui' in b['path']]
    service_files = [b['id'] for b in buildings if 'services' in b['path']]

    # UI -> API
    for uf in ui_files:
        if random.random() > 0.4 and api_files:
            edges.append({"source": uf, "target": random.choice(api_files), "type": "imports"})

    # API -> Services
    for af in api_files:
        if random.random() > 0.3 and service_files:
            edges.append({"source": af, "target": random.choice(service_files), "type": "imports"})

    # Services -> Core
    for sf in service_files:
        if random.random() > 0.3 and core_files:
            edges.append({"source": sf, "target": random.choice(core_files), "type": "imports"})

    # API -> Core
    for af in api_files:
        if random.random() > 0.4 and core_files:
            edges.append({"source": af, "target": random.choice(core_files), "type": "imports"})

    # Random sparse cross-dependencies
    for b in buildings:
        if random.random() > 0.85:
            target = random.choice(buildings)['id']
            if target != b['id']:
                edges.append({"source": b['id'], "target": target, "type": "imports"})

    city_data = {
        "id": "demo_react_app",
        "name": "Codebase City Engine DEMO",
        "path": "demo/react-starter",
        "status": "ready",
        "git_metadata": {
            "branch": "main",
            "last_commit": "abcdef1234567890",
            "commit_msg": "Initial demo city commit"
        },
        "buildings": buildings,
        "roads": edges,
        "metadata": {
            "total_files": len(buildings),
            "total_loc": sum(b['metrics']['loc'] for b in buildings),
            "dominant_language": "javascript",
            "issues": {
                "god_objects": [random.choice(buildings)['id'] for _ in range(3)],
                "circular_dependencies": [[random.choice(buildings)['id'], random.choice(buildings)['id']]]
            }
        }
    }

    # Get the directory of the current script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(script_dir)
    output_path = os.path.join(project_root, 'frontend', 'public', 'demo-city.json')

    with open(output_path, 'w') as f:
        json.dump(city_data, f, indent=2)

    sys.stderr.write(f"✅ Generated demo city with {len(buildings)} buildings and {len(edges)} roads\n")

if __name__ == '__main__':
    generate_demo_city()

File: scripts/generate-demo-city.mjs
/**
 * generate-demo-city.mjs — produces a procedural demo city ~50x larger
 * than the previous static fixture, using the live `buildingTypes` and
 * `colorUtils` so the demo stays in sync with the rendering pipeline.
 *
 * Usage (one-shot):
 *   node scripts/generate-demo-city.mjs
 *
 * Output: /public/demo-city.json
 */

import { writeFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, resolve } from 'node:path'

import { classify, dimsFor, typeNameToId } from '../src/widgets/city-viewport/ui/buildingTypes.js'
import { layoutDistricts } from '../src/widgets/city-viewport/ui/cityLayout.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const OUT = resolve(__dirname, '../public/demo-city.json')

// ── Synthetic file tree ──────────────────────────────────────────────
// 30 directories, each with 100–200 files. ~4500 buildings total.
const DIR_TEMPLATES = [
    { name: 'apps/web/src', files: 220, exts: ['.tsx', '.ts', '.css'] },
    { name: 'apps/api/src/routes', files: 180, exts: ['.ts', '.js'] },
    { name: 'apps/api/src/services', files: 160, exts: ['.ts'] },
    { name: 'apps/api/src/models', files: 130, exts: ['.ts'] },
    { name: 'apps/mobile/src', files: 200, exts: ['.tsx', '.ts'] },
    { name: 'apps/admin/src', files: 140, exts: ['.tsx', '.css'] },
    { name: 'packages/ui/src', files: 180, exts: ['.tsx', '.css'] },
    { name: 'packages/core/src', files: 160, exts: ['.ts'] },
    { name: 'packages/utils/src', files: 110, exts: ['.ts'] },
    { name: 'packages/config/src', files: 90,  exts: ['.ts', '.json', '.yaml'] },
    { name: 'services/auth/src', files: 120, exts: ['.go'] },
    { name: 'services/billing/src', files: 140, exts: ['.go', '.sql'] },
    { name: 'services/notifications/src', files: 100, exts: ['.go'] },
    { name: 'services/analytics/src', files: 130, exts: ['.py'] },
    { name: 'services/search/src', files: 110, exts: ['.rs'] },
    { name: 'services/feed/src', files: 150, exts: ['.go'] },
    { name: 'docs', files: 80, exts: ['.md'] },
    { name: 'scripts', files: 60, exts: ['.sh', '.py'] },
    { name: 'infra/terraform', files: 70, exts: ['.tf', '.yml'] },
    { name: 'infra/k8s', files: 90, exts: ['.yaml'] },
    { name: 'infra/docker', files: 30, exts: ['.dockerfile', '.sh'] },
    { name: 'tests/integration', files: 180, exts: ['.ts', '.js'] },
    { name: 'tests/e2e', files: 120, exts: ['.ts'] },
    { name: 'tests/unit', files: 220, exts: ['.ts', '.js'] },
    { name: 'tools/cli', files: 90, exts: ['.ts', '.go'] },
    { name: 'tools/codegen', files: 70, exts: ['.ts', '.py'] },
    { name: 'libs/database', files: 100, exts: ['.ts', '.sql'] },
    { name: 'libs/cache', files: 80, exts: ['.ts'] },
    { name: 'libs/queue', files: 90, exts: ['.ts', '.go'] },
    { name: 'libs/observability', files: 110, exts: ['.ts', '.go'] },
]

const ROOT_FILES = [
    'README.md', 'LICENSE', 'CONTRIBUTING.md', 'CHANGELOG.md',
    'package.json', 'pnpm-workspace.yaml', '.gitignore',
    'Dockerfile', 'docker-compose.yml', 'tsconfig.json',
    'vite.config.ts', '.eslintrc', '.prettierrc',
]

const AUTHORS = [
    'James Wilson', 'Priya Sharma', 'Diego Garcia', 'Mei Chen', 'Aisha Khan',
    'Liam O\'Connor', 'Sofia Rossi', 'Yuki Tanaka', 'Olu Adeyemi', 'Eva Hansen',
]

const LANG_BY_EXT = {
    '.ts': 'typescript', '.tsx': 'typescript',
    '.js': 'javascript', '.jsx': 'javascript', '.cjs': 'javascript', '.mjs': 'javascript',
    '.py': 'python', '.go': 'go', '.rs': 'rust',
    '.css': 'css', '.scss': 'css', '.html': 'html',
    '.md': 'markdown', '.yml': 'yaml', '.yaml': 'yaml',
    '.json': 'json', '.sh': 'shell', '.sql': 'sql',
    '.tf': 'terraform', '.dockerfile': 'docker',
}

// Deterministic-ish RNG so successive runs produce the same demo.
let _seed = 1234567
function rand() { _seed = (_seed * 9301 + 49297) % 233280; return _seed / 233280 }

function logSize() {
    // log-distributed file size: most small (~1-5 KB), some mid, few big.
    return Math.floor(Math.exp(rand() * 9) + 200)
}

// ── Generate flat file list ──────────────────────────────────────────
const sourceFiles = []
for (const f of ROOT_FILES) {
    sourceFiles.push({ path: f, size: 1500 + Math.floor(rand() * 8000) })
}
for (const tpl of DIR_TEMPLATES) {
    for (let i = 0; i < tpl.files; i++) {
        const ext = tpl.exts[i % tpl.exts.length]
        const size = logSize()
        sourceFiles.push({ path: `${tpl.name}/file_${i}${ext}`, size })
    }
}

// ── Group by directory ───────────────────────────────────────────────
const dirGroups = {}
for (const f of sourceFiles) {
    const dir = f.path.includes('/') ? f.path.substring(0, f.path.lastIndexOf('/')) : '(root)'
    if (!dirGroups[dir]) dirGroups[dir] = []
    dirGroups[dir].push(f)
}

// ── Zoned, contiguous layout — shared with the live analyzer ────────
const placements = layoutDistricts(dirGroups)
const dirNames = placements.map(p => p.dir)
const districtCellSizes = placements.map(p => p.cellSize)

const DISTRICT_COLORS = [
    '#5aa8ff', '#ff8b5a', '#5affb4', '#ffd85a', '#b35aff',
    '#ff5aa0', '#5affff', '#ff5a5a', '#a0ff5a', '#5a8aff',
    '#ff5ae0', '#70ff5a', '#5affaa', '#ff9a5a', '#5accff',
]

// ── Build districts ──────────────────────────────────────────────────
const districts = []
const districtMap = {}
placements.forEach((p, idx) => {
    const districtId = `district_${idx}`
    const half = p.cellSize / 2
    districts.push({
        id: districtId,
        name: p.dir,
        color: DISTRICT_COLORS[idx % DISTRICT_COLORS.length],
        center: { x: p.cx, y: p.cz },
        boundary: [
            { x: p.cx - half, y: p.cz - half },
            { x: p.cx + half, y: p.cz - half },
            { x: p.cx + half, y: p.cz + half },
            { x: p.cx - half, y: p.cz + half },
        ],
        building_count: dirGroups[p.dir].length,
    })
    districtMap[p.dir] = districtId
})

// ── Build buildings ──────────────────────────────────────────────────
const maxFileSize = Math.max(...sourceFiles.map(f => f.size))
const buildings = []
const languageCounts = {}

for (let di = 0; di < dirNames.length; di++) {
    const dir = dirNames[di]
    const files = dirGroups[dir]
    const districtId = districtMap[dir]
    const district = districts[di]
    const cellSize = districtCellSizes[di]
    const bcols = Math.ceil(Math.sqrt(files.length))
    const brows = Math.ceil(files.length / bcols)
    const usableSize = cellSize - 36
    const rawSpacing = usableSize / Math.max(bcols, brows)
    const spacing = Math.max(64, rawSpacing)

    files.forEach((file, i) => {
        const ext = file.path.includes('.') ? file.path.slice(file.path.lastIndexOf('.')) : ''
        const lang = LANG_BY_EXT[ext.toLowerCase()] || 'unknown'
        languageCounts[lang] = (languageCounts[lang] || 0) + 1

        const sizeNorm = Math.log2(file.size + 1) / Math.log2(maxFileSize + 1)
        const buildingType = classify(file, sizeNorm)
        const { width, height, depth } = dimsFor(buildingType, sizeNorm)
        const buildingTypeId = typeNameToId(buildingType)

        const fcol = i % bcols
        const frow = Math.floor(i / bcols)
        const x = district.center.x + (fcol - (bcols - 1) / 2) * spacing
        const z = district.center.y + (frow - (brows - 1) / 2) * spacing

        const name = file.path.includes('/') ? file.path.slice(file.path.lastIndexOf('/') + 1) : file.path
        const author = AUTHORS[Math.floor(rand() * AUTHORS.length)]
        const complexity = Math.ceil(1 + file.size / 500)
        const churn = Math.floor(rand() * 12)

        buildings.push({
            id: file.path,
            name,
            path: file.path,
            file_path: file.path,
            language: lang,
            author,
            email: null,
            district_id: districtId,
            directory: dir,
            position: { x, y: 0, z },
            dimensions: { width, height, depth },
            building_type: buildingType,
            building_type_id: buildingTypeId,
            color_metric: sizeNorm,
            coupling_score: 0,
            lines_of_code: Math.ceil(file.size / 40),
            complexity,
            in_degree: 0,
            functions: [],
            classes: [],
            imports: [],
            metrics: {
                size_bytes: file.size,
                loc: Math.ceil(file.size / 40),
                complexity,
                churn,
                commits: Math.floor(rand() * 30),
                age_days: Math.floor(rand() * 400),
                dependencies_in: 0,
                debt: sizeNorm > 0.7 ? sizeNorm : 0,
            },
        })
    })
}

const cityData = {
    id: 'demo_aurora',
    name: 'Aurora — Demo Metropolis',
    path: 'aurora/metropolis',
    status: 'ready',
    source: 'demo',
    branch: 'main',
    buildings,
    districts,
    metrics: {
        total_files: buildings.length,
        total_lines: buildings.reduce((s, b) => s + (b.metrics?.loc || 0), 0),
        languages: languageCounts,
        contributors: AUTHORS.length,
    },
}

writeFileSync(OUT, JSON.stringify(cityData))
console.log(`Wrote ${buildings.length} buildings across ${districts.length} districts → ${OUT}`)
console.log(`File size: ${(JSON.stringify(cityData).length / 1024 / 1024).toFixed(2)} MB`)
